Travisdouce/do not return active model serializer json

tdouce opened the pull request.

tdouce committed:
Creates feature spec

tdouce committed:
Moves controller test to feature test

This test should live in a feature spec. Controller specs should test
the general logic/flow and not ensuring the JSON is formatted correctly.

tdouce committed:
Generates user resource

tdouce committed:
Creates user resource in db

tdouce committed:
Adds 'active_model_serializers' as a development dependency

tdouce committed:
Creates UserSerializer in dummy application

The feature spec ‘creating_models_remotely’ spec should return the
active_model_serializers serialization for the user resource

tdouce committed:
Stub out FactoryGirl.create

User.create(opts) throw a `ActiveModel::ForbiddenAttributesError`
because ‘first_name’ isn’t permitted with strong_parameters

tdouce committed:
Adds development dependency 'factory_girl_rails'

tdouce committed:
Replace mocked FactoryGirl with "real" FactoryGirl

The problems with stubbing FactoryGirl are:

1. This feature spec should be testing the full stack rather than
mocking/stubbing out interfaces;
2. The mocked FactoryGirl’s .create method created a user resource
using User.create(opts). The problem is that Rails 4 requires that
attributes be whitelisted for mass-assignment with strong parameters. I
didn’t want to implement the Rails 4 strong parameters because it is
possible that some users of this gem are using a version of Rails that
doesn’t support strong parameters. The ‘real’ FactoryGirl avoids this
problem

tdouce committed:
Define factories in spec_hepler :(

For some reason, the factories defined in the dummy app are not
available in the test suite.

tdouce committed:
Require 'active_model_serializers' in feature spec

‘active_model_serializers’ is not loaded until explicitly required. The
test now fails and is returning the active_model_serializer
(UserSerializer) JSON.

tdouce committed:
Returns JSON from #attributes

This is a bit of a hack, but it prevents the app from return
active_model_serializers  JSON.

tdouce committed:
Parse response body

Now that we are creating the user resource with factory_girl, the
`response.body` is returning more than just `{“first_name”:”Sam”}`.
Instead it is returning json like:
`"{\"id\":1,\"first_name\":\"Sam\",\"created_at\":\"2014-11-13T21:42:16.
205Z\",\"updated_at\":\"2014-11-13T21:42:16.205Z\”}”`

Instead of comparing the json as string equality (which is difficult
because of the changing date attributes), we are parsing the JSON and
comparing key/values.

tdouce committed:
Create helper to parse JSON

tdouce committed:
Use named route instead of string literal

tdouce committed:
Rename variable from 'factory' to 'resource'

An ActiveModel is object is returned when `FactoryGirl.create` is
invoked. ‘factory’ implies that a factory_girl object is returned, and
this is misleading.

tdouce committed:
Moves controller spec test to feature spec

This test belongs in a feature spec rather than a controller spec.
Controller specs should be testing the general logic of the controller.
Features specs are more appropriate for comparing JSON, like this test
does

tdouce committed:
Use spec helper to parse json and compare key/values

tdouce committed:
Removes redefining of FactoryGirl from controller spec

tdouce committed:
Amend spec to return a user resource

This spec was failing because #attributes was invoked on `nil` in
homes_controller#create.

tdouce commented:
@cs Here is my take on the solution that does not return JSON serialized with `active_model_serializers`. In short I:
1. Created a `user` resource in the `dummy` application
2. Created a `UserSerializer` in the `dummy` application
3. Installed `factory_girl` rather than defining its interface in the test suite.

Let me know what you think.

tdouce commented on app/controllers/remote_factory_girl_home_rails/home_controller.rb:
```diff
@@ -5,8 +5,8 @@ class HomeController < ApplicationController
-        factory = FactoryGirl.create(factory(params), attributes(params))
-        render json: factory
+        resource = FactoryGirl.create(factory(params), attributes(params))
+        render json: resource.attributes
```
We should probably create a `RemoteFactoryGirlHomeRails::Serializer` class that provides an interface for the serialization. Something like:

``` ruby
module
  class Serializer
    def self.serialize(resource)
       resource.attributes
    end
  end
end
```

This would encourage any users of the gem who want to provide any custom serializations to use this class rather than redefining `#to_json` on all their apps resources/models.

cs commented on remote_factory_girl_home_rails.gemspec:
```diff
@@ -16,4 +16,6 @@ Gem::Specification.new do |s|
 
   s.add_development_dependency "sqlite3"
   s.add_development_dependency 'rspec-rails'
+  s.add_development_dependency 'active_model_serializers'
```
Are you sure that this is still required? If we are really going to stick with `render json: resources.attributes`, `active_model_serializers` cannot interfere with the serialization...

tdouce replied:
You are right. It does seem weird to add a dependency only to ensure in a test that the dependency isn't used. I'll remove it.

cs commented on spec/dummy/db/schema.rb:
```diff
@@ -0,0 +1,22 @@
+
+ActiveRecord::Schema.define(version: 20141113194436) do
+
+  create_table "users", force: true do |t|
```
Maybe we can use `ActiveModel` directly and avoid the database? I'm not 100% sure that's possible, but it would be nice if we could keep things as simple as possible. Simplicity is probably the thing I like the most about `FactoryGirlRails` -- you can read and understand all of its code in a couple of hours.

tdouce replied:
Yes, that is a good thought. My initial implementation did just that. It looked like:

``` ruby
class User
  include ActiveModel::Model
  attr_accessor :first_name, :attributes

  def initialize(attributes={})
    @attributes = attributes
  end
end

class FactoryGirl
  def self.create(factory, opts = {})
    User.new(opts)
  end

  def self.factories
    [
      OpenStruct.new('name' => 'Sam'),
      OpenStruct.new('name' => 'Pete')
    ]
  end
end
```

Ultimately, I think it is more robust test suite to:
1. Add `factory_girl_rails` as a development dependency
2. Create a `user` resource in the dummy app
3. Define a `user` factory in the dummy app

and test the full stack rather than implementing the `User` and `FactoryGirl` interfaces in the test suite. This might add a bit of complexity because someone would have to do a bit more sleuthing to figure out where `User` and `user` factory are defined. However, defining the `User` and `user` factory in the dummy app follows common Rails conventions, so hopefully they will not be too hard to track down.

If this ends up being the wrong decision, we can always refactor:)

tdouce committed:
Removes active_model_serializers

active_model_serializers isn't required for any of the tests to pass.
It was only included to ensure that it was not being used to serialize
resources.

tdouce committed:
Load factories defined in dummy app

tdouce committed:
Adds Serializer

This class acts as a facade for implementing #attributes on a resource.
The purpose of this class is to encourage any users of the gem who want
to provide any custom serializations to use this class rather than
redefining #to_json on all their apps resources/models.

tdouce commented:
@cs Is there anything we've missed/need to add before we merge this?

cs commented on spec/controllers/home_controller_spec.rb:
```diff
@@ -35,14 +22,9 @@ def self.factories
 
       it 'should create a factory' do
-        expect(FactoryGirl).to receive(:create).with(:user, {})
+        expect(FactoryGirl).to receive(:create).with(:user, {}).and_return(FactoryGirl.create(:user))
```
This works?? :) Really?

cs replied:
I mean, you are calling exactly the method you've just mocked before (i.e. `#create`). The only reason why it works, is the extra argument in the `#with` ...

I suppose it's fine for now. However, as you know, things like that tend to break easily :)

tdouce replied:
:) Nice catch. Not sure what I had in mind.

cs commented:
Everything is looking good for me. Let's do it! :+1:

cs commented:
Once this is merged, could you please release a new version of the Gem too?

tdouce commented:
re: new version of Gem.

Yep. Thanks for your hard work!

tdouce committed:
Stub Factory.create with user resource

Instead of stubbing FactoryGirl.create with “FactoryGirl.create(:user).
A user resource should be returned when FactoryGirl.create is invoked.
It is odd that the object that is returned from the stub is the same
object that is being stubbed.

tdouce committed:
Reword spec to be more specific

Specify what is being created

tdouce commented:
@cs I released a [new version 1.0.1](https://rubygems.org/gems/remote_factory_girl_home_rails)

cs commented:
@tdouce Thank you!
